### PR TITLE
Netty refactor

### PR DESCRIPTION
**Pull request by kamcio96**
Done:
- move compress package
- close ChannelWrapper
- add ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE to write method
- bossEventLoopGroup, workerEventLoopGroup

@md-5 I hope that will react to this :wink:

### CODE REVIEW

File: proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java [hunk 1]
```diff
@@ -38,30 +35,44 @@ public void setVersion(int protocol)
...
             if ( packet instanceof PacketWrapper )
             {
                 ( (PacketWrapper) packet ).setReleased( true );
-                ch.write( ( (PacketWrapper) packet ).buf, ch.voidPromise() );
+                ch.write( ( (PacketWrapper) packet ).buf ).addListener( ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE );
             } else
             {
-                ch.write( packet, ch.voidPromise() );
+                ch.write( packet ).addListener( ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE );
             }
```
Comment (minecrafter) [inline proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java]: Again, you should really use `voidPromise()` as it as the same effect without torturing the garbage collector,

Comment (kamcio96) [inline proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java]: No, voidPromise() doesn't throw errors. I checked it few times.

Comment (kamcio96) [inline proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java]: @minecrafter are you sure void promise will throw exception?

**Comment by minecrafter** [comment]
@kamcio96 I think I have already explained this to you prior. `FIRE_EXCEPTION_ON_FAILURE` calls `exceptionCaught()`, and `Channel.voidPromise()` does the same thing.

**Comment by Techcable** [comment]
I love this change :)
It ensures packets reach the client before closing the channel, which is a known issue (we already have a cludgy workaround).
It also uses the best practice of seperate worker and boss event groups.

**Comment by PunKeel** [comment]
up? :)

**Comment by kamcio96** [comment]
refresh

**Comment by md-5** [comment]
You have four basically unrelated changes with absolutely no indication of what is meant to be addressed....

File: proxy/src/main/java/net/md_5/bungee/BungeeCord.java [hunk 2]
```diff
@@ -71,7 +71,7 @@
 import net.md_5.bungee.api.plugin.Plugin;
 import net.md_5.bungee.api.plugin.PluginManager;
 import net.md_5.bungee.command.*;
-import net.md_5.bungee.compress.CompressFactory;
+import net.md_5.bungee.netty.compress.CompressFactory;
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/BungeeCord.java]: Absolutely no reason for this

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 3]
```diff
@@ -219,7 +217,7 @@ public void done(ServerPing result, Throwable error)
                     public void done(ProxyPingEvent pingResult, Throwable error)
                     {
                         Gson gson = BungeeCord.getInstance().gson;
-                        unsafe.sendPacket( new StatusResponse( gson.toJson( pingResult.getResponse() ) ) );
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Why

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 4]
```diff
@@ -511,7 +508,13 @@ public void run()
                             thisState = State.FINISHED;
                         }
                     }
-                } );
+                };
+
+                if ( ch.getHandle().eventLoop().inEventLoop() ) {
+                    nettyPostCall.run();
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: Why

File: proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java [hunk 5]
```diff
@@ -76,6 +76,7 @@ public void disconnected(ChannelWrapper channel) throws Exception
             {
                 player.unsafe().sendPacket( packet );
             }
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java]: Why

File: proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java [hunk 6]
```diff
@@ -1,12 +1,10 @@
 package net.md_5.bungee.netty;
 
-import net.md_5.bungee.compress.PacketCompressor;
-import net.md_5.bungee.compress.PacketDecompressor;
+import io.netty.channel.*;
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java]: No

File: proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java [hunk 7]
```diff
@@ -54,14 +51,28 @@ public void write(Object packet)
...
             ch.close();
         }
     }
 
+    public void close(Object packet)
+    {
+        if ( !isClosed() )
+        {
+            closed = true;
+            ch.writeAndFlush( packet ).addListener( ChannelFutureListener.CLOSE );
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java]: Does not necessarily fix the issues described

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java [hunk 8]
```diff
@@ -531,24 +534,14 @@ public void disconnect(final BaseComponent... reason)
...
-                public void run()
-                {
-                    if ( thisState != State.STATUS && thisState != State.PING )
-                    {
-                        unsafe().sendPacket( new Kick( ComponentSerializer.toString( reason ) ) );
-                    }
-                    ch.close();
-                }
-            }, 500, TimeUnit.MILLISECONDS );
+            if ( thisState != State.STATUS && thisState != State.PING )
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: This reintroduces the severity of race condition above

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 4 above)
Comment (kamcio96) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: If plugin doesn't change thread we can execute it immediately. Netty doesn't do it.

File: proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java (hunk 5 above)
Comment (kamcio96) [inline proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java]: Client close connection -> bungee close connection with server -> bungee will try kick player

File: proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java (hunk 7 above)
Comment (kamcio96) [inline proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java]: I tested it for long time. Eg?

File: proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java (hunk 3 above)
Comment (kamcio96) [inline proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java]: unsafe is for plugins. Theory it does not matter. But its implementation, and I decided to use channel class

**Pull request closed without merging**